engine/src/main/java/org/hibernate/search/engine/mapper/mapping/building/impl/IndexSchemaFilter.java

```diff
@@ -196,24 +206,91 @@ private String getPathFromSameIndexedEmbeddedSinceNoCompositionLimits(IndexedEmb
+	 * - we want to know the filter representing the definition `c.` (marked with `^`) that is right before the `a.` with index `2` as well as the
+	 * cycle-path leading to it from `5` -- "a.b.c.a.b.c.a.b.c."
+	 * that will be the filter and the path we'll pass to the "isPotentiallyExcludedPath" check. These will be our `previousCycleFilter` and `previousCycle`
+	 *
+	 * Doing so will go through any potential exclude filters defined on indexed-embedded only once; going from that definition `c.` (^)
+	 * and through its parents will cover the definitions from the cycle itself as well as any definitions out of the cycle that are in the path leading to the cycle.
+	 *
+	 * In case we don't have a repeating cycle yet - we'll be starting from the "current filter"
+	 * (previousCycleFilter == this, previousCycle == this.definition.relativePrefix() )
+	 */
+	private String findCycle(IndexedEmbeddedDefinition definitionToBeAdded) {
+		String shortestCycle = null;
+		String currentCycle = null;
+		String previousCycle = this.definition.relativePrefix();
+		IndexSchemaFilter localParent = this;
+		IndexSchemaFilter currentCycleFilter = null;
+		IndexSchemaFilter previousCycleFilter = this;
+		StringBuilder path = new StringBuilder();
+		while ( localParent != null && localParent.definition != null ) {
+
+			path.insert( 0, localParent.definition.relativePrefix() );
+
+			if ( localParent.definition.equals( definitionToBeAdded ) ) {
+
+				if ( currentCycleFilter != null ) {
+					previousCycleFilter = currentCycleFilter.parent;
+					previousCycle = currentCycleFilter.definition.relativePrefix() + currentCycle;
+				}
+
+				currentCycle = path.toString();
+				currentCycleFilter = localParent;
+
+				if ( shortestCycle == null ) {
+					// Set the first found cycle and keep it unchanged till the end of the loop. We'll report it to the user.
+					shortestCycle = currentCycle;
+				}
 			}
+
+			localParent = localParent.parent;
+		}
+
+		if ( currentCycle == null ) {
+			return null;
 		}
-		else {
-			/*
-			 * No recursion limits, no parent: this is the root.
-			 * I we reach this point, it means there was no recursion limit at all,
-			 * but we did not encounter the IndexedEmbedded we were looking for.
-			 */
+
+		// If we reached this point -- we've found a cycle, and we need to check if any exclude paths can break it:
+		if ( isPotentiallyExcludedPath(
+				// when we check the node we don't need the node's relative path at the beginning of the path we are checking:
+				previousCycle.substring( previousCycleFilter.definition.relativePrefix().length() ),
+				previousCycleFilter
+		) ) {
 			return null;
 		}
+
+		// We haven't found an exclude that would break the cycle - so we return the shortest cycle
+		return shortestCycle;
+	}
+
+	private String removeDot(String string) {
+		return string != null && string.endsWith( "." ) ? string.substring( 0, string.length() - 1 ) : string;
+	}
```

Comment:
Didn't we discuss just ignoring that kind of things? It didn't work?

Reply:
mmm, I think we've talked about ignoring this in the `PathFilter#isPotentiallyExcluded()` where we wouldn't care if we may cut some property into two parts and would just use a "starts with"... as eventually the cycle will grow, and the path will not be "potentially excluded" anymore.
Here though, we need to pass something to the `IndexSchemaFilter`, and that one expects that there are no dots in the end... So I guess we need this one, but calling it inside of the recursion makes no sense -- I'll move the call before the recursion happens.

Reply:
Ok. In that case... if you end up with a trailing dot, you're not manipulating a path... there's some trailing prefix that you probably should not even have included in the first place. 

Maybe instead of removing a trailing dot, you should be removing everything after (and including) the last dot at some point? I think we even discussed that...

Reply:
Bump, just in case you missed this :)


> Ok. In that case... if you end up with a trailing dot, you're not manipulating a path... there's some trailing prefix that you probably should not even have included in the first place.
> 
> Maybe instead of removing a trailing dot, you should be removing everything after (and including) the last dot at some point? I think we even discussed that...

Reply:
> Let's take a simple case where we don't have any prefixes messing up things for us. And we have a cycle:
> [...]
> Our exclude paths in the filters do not have . at the end of the path so passing pathFromPreviousCycleFilter as is won't work. And in this case, removing the dot at the end of the string is what I'd expect to happen.

Agreed.

> now a case with a prefix-without-a-dot. Let's say we have :
> [...]
> So at this point, we can either send c.a.prefix_c.a.prefix_c.a.prefix_ (path as is) or c.a.prefix_c.a.prefix_c.a (path without an "unterminated prefix" in the end that's if I understand you correctly is what you are suggesting) and both would work fine with our path filters

Agreed.

> now if we have a case without any dots in the prefixes....
> [...]
> then we will only have an infinite sequence of prefix_a_prefix_b_+, and we won't be able to pass anything to our path filter if we'd only wanted to pass whatever comes prior to the last dot. so that would lead to an overflow and cycle not being detected ...

In this particular case, we know there's an infinite cycle. `excludePaths` don't matter as they cannot break the cycle (since they will never match any path in the cycle). So IMO it's fine if either:

* `isPotentiallyExcluded()` returns `false` for an empty path
* or if you just add an `if` in `findCycle` that doesn't even call `isPotentiallyExcluded()` if "pathFromPreviousFilterAfterRemovingTrailingPrefix" is empty.